census2020/census2020.ipynb

```diff
@@ -0,0 +1,609 @@
+    "    return hex_colors"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "races = [\"All\", *df['race'].unique().compute()]\n",
+    "\n",
+    "class USPopulation(pn.viewable.Viewer):\n",
+    "    race      = param.Selector(default='All', objects=races)\n",
+    "    city      = param.Selector(default='USA', objects=list(CITIES.keys()))\n",
+    "    map_tile  = param.Boolean(default=False, doc=\"Turn the map tiles on/off\")\n",
+    "    map_alpha = param.Magnitude(default=0.5, doc=\"Alpha value for the map opacity\")\n",
+    "    \n",
+    "    points = param.Parameter()\n",
+    "    base_map = param.Parameter()\n",
+    "    plot = param.Parameter()\n",
+    "\n",
+    "    \n",
+    "    def __init__(self, **params):\n",
+    "        super().__init__(**params)\n",
+    "        self.points = None\n",
+    "        self.base_map = None\n",
+    "        self._update_data_points()\n",
+    "        self._update_map()\n",
+    "\n",
+    "    def get_data(self, race):\n",
+    "        if race == 'All':\n",
+    "            return df\n",
+    "        else:\n",
+    "            return df[df['race'] == race]\n",
+    "    \n",
+    "    @pn.cache\n",
+    "    def get_colormap(self, race):\n",
+    "        if race == 'All':\n",
+    "            return color_key\n",
+    "        else:\n",
+    "            color = color_key.get(race, 'aqua')\n",
+    "            return create_colormap_hex_list(color)\n",
+    "    \n",
+    "    @param.depends('race', 'city', watch=True)\n",
+    "    def _update_data_points(self):\n",
+    "        \"\"\"Update the data points based on race and city.\"\"\"\n",
+    "        df = self.get_data(self.race)\n",
+    "        cmap = self.get_colormap(self.race)\n",
+    "        x_range, y_range = CITIES[self.city]\n",
+    "        \n",
+    "        plot_options = {\n",
+    "            'x': 'x',\n",
+    "            'y': 'y',\n",
+    "            'rasterize': True,\n",
+    "            'dynspread': True,\n",
+    "            'cnorm': 'eq_hist',\n",
+    "            'xlim': x_range,\n",
+    "            'ylim': y_range,\n",
+    "            'cmap': cmap,\n",
+    "            'bgcolor': 'black',\n",
+    "            'colorbar': False,\n",
+    "            'hover': False,\n",
+    "            'aggregator': ds.by('race') if self.race == 'All' else None,\n",
+    "            'width': plot_width,\n",
+    "            'height': plot_height,  \n",
+    "        }\n",
+    "\n",
+    "        \n",
+    "        self.points = df.hvplot.points(**plot_options) # * label (causes the plot to go all white)\n",
+    "        \n",
+    "        if self.race == 'All':\n",
+    "            self.points *= legend\n",
+    "        \n",
+    "        self._update_plot()\n",
+    "        \n",
+    "    @param.depends('map_tile', 'map_alpha', 'city', watch=True)\n",
```

Comment:
Agreed. The `datashader_dashboard` code shows how to avoid that update hit for alpha (e.g. using .apply.opts) and for tiles (which would be done as a separate callback, not folded into a single one.  I'm only seeing these notes here; I'm not sure what this app is, as the original census doesn't have an app.

Reply:
For plots that are quick to render, I don't mind a Panel-based approach that just re-renders the plot when a widget is updated, as this API (Panel's) is a bit easier to apprehend. In this case, we're rendering a large dataset and should indeed be a bit more careful when updating the plot, relying on HoloViews' API more.